### ISSUE TITLE
Large server list issues

### ISSUE THREAD

**Issue by limejuce** [Austinb/GameQ#416]
I have database of 130k servers. Tried many different ways to achieve the best results but none works as expected. 

PHP version 7.2, tried on 5.6, 7.1 as well.
Web hosting details:
Dedicated
Proccessor - 9.60 GHz (4 x 2.40 GHz)
Memory - 8192 MB + 896 MB extra + 4096 MB SWAP
Network - 1000 Mbps

Contacted my hosting support, to check if ports are not blocked and as expected everything is fine. Bzip2 compression enabled as well. 

So basically what happens. If I try to process more then 20 servers at once. Nearly all of servers returns with NULL variables. Basically out of few test from 130k servers only received up to 500 online servers with correct details. 

Did a test where I array_chunk'ed the whole array of servers into 20 per go, most of servers returned online. But process was so slow. To query all of servers would take half of day (20 servers every 1-2 seconds and total 130k servers to check). 

Tried many different variations of 'advanced settings'. Had time_out setting from default to 20 seconds. Stream_select also been tested on up to 10 seconds. Write_wait from 0 to 1500. Many different combinations. Had a try with all of these on default.

Did a test where I setted up to query all 130k servers in one go, had time_out - 20, write_wait - 1000, stream_select - 1500000. The whole array was processed in less then 8 minutes. But nearly all of servers returned invalid data - NULL. 

Tried while function for each server,  scanned each server 3 times until code decides that server is off. 

Even tried to query with GameQ and if server returns invalid data to process it with lgsl library. Had much more servers queried as online. But lgsl library was too slow.

Also had a test where each server I processed individually, with addServer. Querying was too slow. Noticed better results in speed when querying more.

I am willing to donate for decent solution. As project is ready to go, just this querying mechanism is left.

**Comment by thereis** [comment]
Get out from PHP and try NodeJS. You can do async tasks more easier thank php and create parallel tasks for each server execution.

Node can handle this easily. A node package: [Gamedig](https://www.npmjs.com/package/gamedig)

Just create a simple rest service with Express and pass as argument the ip and port to your rest.

**Comment by limejuce** [comment]
Thank you for reply, Reis. Never used NodeJS. Is there any chance to get short consultation through the email?

**Comment by thereis** [comment]
You can see my e-mail in my profile. Feel free to keep in contact.

**Comment by slowerua** [comment]
I have the same problem.
**Fix:**
Changed GameQ version in composer.json and it works (tested with list from 20 different servers).
    "require": {
        "austinb/gameq": "v3.0.11"
    }

**Comment by limejuce** [comment]
Issue is still not solved

**Comment by Austinb** [comment]
Someone had posted about rolling back to v3.0.11 and see if that helped.  Have you tried that? It is possible the changes in the handling of the sockets broke something with large lists.

**Comment by Austinb** [comment]
Apparently the merge from 385cfb2 broke some of the querying fail safes.  If it ain't broke don't fix it I suppose.  Will push out a new release shortly.  Please test that and let me know if it does not fix your issue.

**Comment by limejuce** [comment]
Querying script is currently running. I can confirm that I receive much more servers as ONLINE, with correct details. Speed is not as fast as I would need, but that might be my script error. Will edit this message tomorrow, when I will spend more time in analyzing all possible ways. But for a fact downgrade made a bigger change on getting more correct results.

EDIT. By the time I was writing this post, 1500+ servers were scanned as online. And settings were:

$GameQ->setOption('timeout', 3);
$GameQ->setOption('stream_timeout', 500000); 
$GameQ->setOption('write_wait', 0); 

As I mentioned before. Tomorrow I will play a little bit more with settings and script. So I will come back with final result. But for a fact it's newer version problem.

EDIT EDIT: cuz of downgrade minecraftpe servers thows exception Uncaught exception 'GameQ\Exception\Server' with message 'Unable to locate Protocols class for 'minecraftpe'!'

**Comment by Austinb** [comment]
Please pull in v3.0.13 (contains query change back to v3.0.11) and let me know if that helps.  There are other bug fixes for PE specifically that should be included in that release.

**Comment by limejuce** [comment]
Results are much better. Basically array_chunck'ed the whole array in 5000 per go. Got 1500 online servers, and the whole list was queried in 1.5minutes. But settings were low. Will give a try with increased 'advanced' settings. Will edit this post later.

**Comment by Austinb** [comment]
Glad to hear it.  Just so you know you will probably have to break up your list into smaller chunks than 5000.  I have not done any real testing as I do not have access to a large list of servers.  If you find a good configuration for querying a large list of servers please let me know. I would like to write a wiki page using what you find if you are okay with that.

**Comment by limejuce** [comment]
Sure, currently doing some test. Will post results tomorrow. At the moment I did 6 different test, with different results. Managed to get 8k+ online servers in 8 minutes. Will post update tomorrow.

**Comment by limejuce** [comment]
Okay....
So here I am back. Today spend couple hours on tests with different configurations and methods.
First thing, tried to add all 130k servers with $GameQ->addServers. Web server crashed couple times, tried with 30k, 10k, 1k no crash, but nearly all servers returns OFFLINE. Let's say out of 10k or 1k as online returns only 100-300 servers, but processing speed is extra fast, doesn't even take 5 minutes to process big amount of servers and doesn't matter what settings you use (believe me, tried probably all possible combinations) querying going to fail.

Soooo, what to do? I found that the best results I get if I send less then 300 servers at once to addServers. Tried with 500 servers per go, pretty fast querying but low success rate on receiving proper details from server. Same with 400, 350, 300. And finally 250, will explain what you will receive if you will send 250 servers.
Limited database to 25k, array pushed it, then array_chunked whole array to 250 servers per go and got 13k-13.5k online servers in 5 minutes. Not bad? Oh and also settings I used:

$GameQ->addFilter('normalise'); 
$GameQ->addFilter('secondstohuman');
$GameQ->setOption('timeout', 3);
//$GameQ->setOption('stream_timeout', 200000);
$GameQ->setOption('write_wait', 0); 
$GameQ->addServers(YOUR ARRAY);

What I did next - I lowered number to 200 servers. AND results are much much better, also have tried little bit different settings. So, results were 18032 servers in 5.98 (nearly 6 minutes).
Settings (only increased timeout a little bit):
$GameQ->addFilter('normalise'); 
$GameQ->addFilter('secondstohuman');
$GameQ->setOption('timeout', 5);
//$GameQ->setOption('stream_timeout', 200000);
$GameQ->setOption('write_wait', 0); 

That's currently the best result I got. Server didn't crashed, got 18k online servers out of 25k total queried. And everything took only 6 minutes. Of course, that's not what I need, but I will carry on doing some experiments until I will come up with the best combination and method.

**Comment by Austinb** [comment]
There has always been a limit to GameQ mainly due to the builtin socket handling of PHP and the need to at some point stop waiting for responses.  It is possible a different socket module would be better and faster. There was an extension for PHP that you had to compile in that showed promise but I can not seem to find it anywhere.  When version 3 was written I actually built the Query(ing) side to be interchangeable.  The native one uses all built in functions from PHP that do not require any extra work.  It may be advantageous to try to bring in another type of sockets library that is built for speed but I can not seem to find that older extension (may not exist anymore).  

If people have ideas on using a third party library or something that has to be compiled into php that will be faster I would be willing to make sure it links up to the querying correctly.  I just do not have any ideas on any other methods for making the calls faster.   Sockets/streams are not my strong suit.  Last change to the sockets that got merged in ended up making things worse. 

Let me know if you find a good combo with the current setup.  I know other people have threaded the library to make calls faster but I am not sure how much that helps or how much work it would take to bring in some kind of threading to this library.
